File: src/vetscribe/api_client.py
import logging
from dataclasses import dataclass

import httpx

logger = logging.getLogger("vetscribe.api_client")


class ApiClientError(Exception):
    pass


@dataclass
class SoapNote:
    subjective: str
    objective: str
    assessment: str
    plan: str
    transcript: str = ""


@dataclass
class Exam:
    """A persisted exam as served by the backend history endpoints — the shared
    record the mobile and desktop apps both read from and edit."""

    id: str
    created_at: str
    subjective: str
    objective: str
    assessment: str
    plan: str
    transcript: str = ""
    patient_name: str | None = None

    @property
    def soap_text(self) -> str:
        # Local import avoids a circular import (pipeline imports api_client).
        from vetscribe.pipeline import format_soap_text

        return format_soap_text(self)

    @classmethod
    def from_dict(cls, data: dict) -> "Exam":
        return cls(
            id=data["id"],
            created_at=data["created_at"],
            subjective=data["subjective"],
            objective=data["objective"],
            assessment=data["assessment"],
            plan=data["plan"],
            transcript=data.get("transcript", ""),
            patient_name=data.get("patient_name"),
        )


class ApiClient:
    def __init__(self, endpoint: str, timeout_seconds: float, api_key: str = ""):
        self.endpoint = endpoint
        self.timeout_seconds = timeout_seconds
        self.api_key = api_key

    def _auth_headers(self, extra: dict) -> dict:
        headers = dict(extra)
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"
        return headers

    def _parse_soap_response(self, response) -> SoapNote:
        logger.info("backend responded with status %s", response.status_code)
        if response.status_code != 200:
            raise ApiClientError(f"backend returned {response.status_code}: {response.text}")

        data = response.json()
        try:
            return SoapNote(
                subjective=data["subjective"],
                objective=data["objective"],
                assessment=data["assessment"],
                plan=data["plan"],
                transcript=data.get("transcript", ""),
            )
        except KeyError as exc:
            raise ApiClientError(f"backend response missing field: {exc}") from exc

    def generate_soap_note(self, audio_bytes: bytes) -> SoapNote:
        headers = self._auth_headers({"Content-Type": "audio/wav"})

        logger.info("POST %s (timeout=%ss)", self.endpoint, self.timeout_seconds)
        try:
            response = httpx.post(
                self.endpoint,
                content=audio_bytes,
                headers=headers,
                timeout=self.timeout_seconds,
            )
        except httpx.RequestError as exc:
            logger.error("backend request failed: %s", exc)
            raise ApiClientError(f"request failed: {exc}") from exc

        return self._parse_soap_response(response)

    def fetch_history(self) -> list:
        # Read the shared exam history from the backend (single source of truth
        # across mobile + desktop). Returns newest-first Exam objects.
        url = f"{self._api_base()}/history"
        headers = self._auth_headers({})
        try:
            response = httpx.get(url, headers=headers, timeout=self.timeout_seconds)
        except httpx.RequestError as exc:
            logger.error("fetch history failed: %s", exc)
            raise ApiClientError(f"request failed: {exc}") from exc

        if response.status_code != 200:
            raise ApiClientError(f"backend returned {response.status_code}: {response.text}")
        return [Exam.from_dict(item) for item in response.json().get("exams", [])]

    def update_exam(
        self,
        exam_id: str,
        *,
        subjective: str,
        objective: str,
        assessment: str,
        plan: str,
        transcript: str,
    ) -> Exam:
        # Save the vet's structured edits back to the authoritative record so
        # they sync to every device.
        url = f"{self._api_base()}/exams/{exam_id}"
        headers = self._auth_headers({"Content-Type": "application/json"})
        payload = {
            "subjective": subjective,
            "objective": objective,
            "assessment": assessment,
            "plan": plan,
            "transcript": transcript,
        }
        try:
            response = httpx.put(url, json=payload, headers=headers, timeout=self.timeout_seconds)
        except httpx.RequestError as exc:
            logger.error("update exam failed: %s", exc)
            raise ApiClientError(f"request failed: {exc}") from exc

        if response.status_code != 200:
            raise ApiClientError(f"backend returned {response.status_code}: {response.text}")
        return Exam.from_dict(response.json())

    def delete_exam(self, exam_id: str) -> None:
        # Remove an exam from the shared history so the deletion syncs everywhere.
        url = f"{self._api_base()}/exams/{exam_id}"
        headers = self._auth_headers({})
        try:
            response = httpx.delete(url, headers=headers, timeout=self.timeout_seconds)
        except httpx.RequestError as exc:
            logger.error("delete exam failed: %s", exc)
            raise ApiClientError(f"request failed: {exc}") from exc

        if response.status_code != 200:
            raise ApiClientError(f"backend returned {response.status_code}: {response.text}")

    def _api_base(self) -> str:
        # The configured endpoint is the audio POST URL (".../api/soap"); the
        # injection endpoints live alongside it under ".../api/injections", so
        # drop the trailing path segment. Keeps a single configured URL.
        return self.endpoint.rstrip("/").rsplit("/", 1)[0]

    def fetch_pending_injections(self) -> list:
        # Poll the backend for mobile-originated requests to paste a note into
        # AVImark. Each request carries the exam's current note to inject.
        url = f"{self._api_base()}/injections/pending"
        headers = self._auth_headers({})
        try:
            response = httpx.get(url, headers=headers, timeout=self.timeout_seconds)
        except httpx.RequestError as exc:
            logger.error("fetch pending injections failed: %s", exc)
            raise ApiClientError(f"request failed: {exc}") from exc

        if response.status_code != 200:
            raise ApiClientError(
                f"backend returned {response.status_code}: {response.text}"
            )
        return response.json().get("requests", [])

    def ack_injection(self, request_id: str, outcome: str) -> None:
        # Tell the backend a pending injection request has been handled so it's
        # dropped from the pending list and never re-delivered.
        url = f"{self._api_base()}/injections/{request_id}/ack"
        headers = self._auth_headers({"Content-Type": "application/json"})
        try:
            response = httpx.post(
                url, json={"outcome": outcome}, headers=headers, timeout=self.timeout_seconds
            )
        except httpx.RequestError as exc:
            logger.error("ack injection failed: %s", exc)
            raise ApiClientError(f"request failed: {exc}") from exc

        if response.status_code != 200:
            raise ApiClientError(
                f"backend returned {response.status_code}: {response.text}"
            )

    def regenerate_soap_note(self, transcript: str) -> SoapNote:
        # Ask the backend to re-run only note generation over a hand-corrected
        # transcript. Derived from the same base endpoint the audio POST uses so
        # only one URL needs configuring.
        url = self.endpoint.rstrip("/") + "/regenerate"
        headers = self._auth_headers({"Content-Type": "application/json"})

        logger.info("POST %s (timeout=%ss)", url, self.timeout_seconds)
        try:
            response = httpx.post(
                url,
                json={"transcript": transcript},
                headers=headers,
                timeout=self.timeout_seconds,
            )
        except httpx.RequestError as exc:
            logger.error("backend regenerate request failed: %s", exc)
            raise ApiClientError(f"request failed: {exc}") from exc

        return self._parse_soap_response(response)

File: mobile/src/services/api/ApiClient.ts
import type { AudioBody, Exam, InjectionRequest, SoapNote } from './types';

export class ApiClientError extends Error {}

/** Minimal Response shape we depend on — decouples us from the DOM/undici lib. */
export interface HttpResponse {
  ok: boolean;
  status: number;
  json(): Promise<unknown>;
  text(): Promise<string>;
}

export interface HttpRequestInit {
  method?: string;
  headers?: Record<string, string>;
  body?: unknown;
  signal?: unknown;
}

export type HttpFetch = (url: string, init?: HttpRequestInit) => Promise<HttpResponse>;

export interface ApiClientOptions {
  baseUrl: string;
  apiKey?: string;
  /** Injectable fetch for tests; defaults to the platform global fetch. */
  fetch?: HttpFetch;
  /** Per-request timeout in ms; 0 disables it. */
  timeoutMs?: number;
}

interface RawExam {
  id: string;
  created_at: string;
  patient_name?: string | null;
  subjective: string;
  objective: string;
  assessment: string;
  plan: string;
  transcript: string;
}

interface RawInjectionRequest {
  id: string;
  exam_id: string;
  created_at: string;
  status: string;
  outcome: string | null;
}

/**
 * REST + sync client for the VetScribe backend. Mirrors the desktop ApiClient's
 * contract (POST /api/soap with raw audio bytes, POST /api/soap/regenerate with a
 * transcript) and adds the cross-device sync surface: /api/history and /api/exams/:id.
 *
 * The backend is the authoritative source of truth for medical records; this client
 * only reads and writes through it — it keeps no local copy.
 */
export class ApiClient {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly fetch: HttpFetch;
  private readonly timeoutMs: number;

  constructor(options: ApiClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/+$/, '');
    this.apiKey = options.apiKey ?? '';
    this.fetch = options.fetch ?? ((globalThis as { fetch?: HttpFetch }).fetch as HttpFetch);
    this.timeoutMs = options.timeoutMs ?? 60000;
  }

  /** Upload a recording and get back the persisted, generated exam note. */
  async generateNote(audio: AudioBody, mimeType: string): Promise<Exam> {
    const raw = await this.request('/api/soap', {
      method: 'POST',
      headers: { 'Content-Type': mimeType },
      body: audio,
    });
    return this.toExam(raw as RawExam);
  }

  /** Fetch all exams (newest first), synced across mobile and desktop sessions. */
  async fetchHistory(): Promise<Exam[]> {
    const raw = (await this.request('/api/history', { method: 'GET' })) as { exams?: RawExam[] };
    return (raw.exams ?? []).map((e) => this.toExam(e));
  }

  /** Fetch a single exam by id. */
  async getExam(id: string): Promise<Exam> {
    const raw = await this.request(`/api/exams/${encodeURIComponent(id)}`, { method: 'GET' });
    return this.toExam(raw as RawExam);
  }

  /** Save the vet's inline edits back to the authoritative backend. */
  async updateExam(id: string, note: SoapNote): Promise<Exam> {
    const raw = await this.request(`/api/exams/${encodeURIComponent(id)}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(note),
    });
    return this.toExam(raw as RawExam);
  }

  /** Delete an exam from the shared history so it's gone on every device. */
  async deleteExam(id: string): Promise<void> {
    await this.request(`/api/exams/${encodeURIComponent(id)}`, { method: 'DELETE' });
  }

  /** Re-run note generation from a hand-corrected transcript (no re-transcription). */
  async regenerateNote(transcript: string): Promise<SoapNote> {
    const raw = (await this.request('/api/soap/regenerate', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ transcript }),
    })) as Partial<RawExam>;
    return {
      subjective: raw.subjective ?? '',
      objective: raw.objective ?? '',
      assessment: raw.assessment ?? '',
      plan: raw.plan ?? '',
      transcript: raw.transcript ?? transcript,
    };
  }

  /**
   * Ask the backend to have this exam's note pasted into AVImark on the desktop.
   * The phone and the exam-room PC never talk directly: the backend queues the
   * request and the desktop tray app polls for it, so this works regardless of
   * which network the phone is on. Returns immediately once queued (HTTP 202) —
   * the actual paste happens on the desktop moments later.
   */
  async requestInjection(examId: string): Promise<InjectionRequest> {
    const raw = (await this.request(`/api/exams/${encodeURIComponent(examId)}/inject`, {
      method: 'POST',
    })) as RawInjectionRequest;
    return {
      id: raw.id,
      examId: raw.exam_id,
      createdAt: raw.created_at,
      status: raw.status,
      outcome: raw.outcome,
    };
  }

  private authHeaders(extra: Record<string, string> = {}): Record<string, string> {
    const headers = { ...extra };
    if (this.apiKey) {
      headers.Authorization = `Bearer ${this.apiKey}`;
    }
    return headers;
  }

  private async request(path: string, init: HttpRequestInit): Promise<unknown> {
    const url = `${this.baseUrl}${path}`;
    const controller = this.timeoutMs > 0 ? new AbortController() : null;
    const timer = controller
      ? setTimeout(() => controller.abort(), this.timeoutMs)
      : null;

    let response: HttpResponse;
    try {
      response = await this.fetch(url, {
        ...init,
        headers: this.authHeaders(init.headers),
        signal: controller?.signal,
      });
    } catch (error) {
      throw new ApiClientError(`request to ${path} failed: ${(error as Error).message}`);
    } finally {
      if (timer) clearTimeout(timer);
    }

    if (!response.ok) {
      const detail = await response.text().catch(() => '');
      throw new ApiClientError(`backend returned ${response.status}: ${detail}`);
    }

    try {
      return await response.json();
    } catch (error) {
      throw new ApiClientError(`invalid JSON from ${path}: ${(error as Error).message}`);
    }
  }

  private toExam(raw: RawExam): Exam {
    return {
      id: raw.id,
      createdAt: raw.created_at,
      patientName: raw.patient_name ?? undefined,
      subjective: raw.subjective,
      objective: raw.objective,
      assessment: raw.assessment,
      plan: raw.plan,
      transcript: raw.transcript,
    };
  }
}
